trin-core/src/types/block_body.rs

```diff
@@ -0,0 +1,307 @@
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use bytes::Bytes;
+use eth_trie::{EthTrie, MemoryDB, Trie};
+use ethereum_types::{Address, H160, H256, U256, U64};
+use rlp::{Decodable, DecoderError, Encodable, Rlp, RlpStream};
+use rlp_derive::{RlpDecodable, RlpEncodable};
+use sha3::{Digest, Keccak256};
+
+use super::{header::Header, receipts::TransactionId};
+
+#[derive(Clone, Debug)]
+pub struct BlockBody {
+    _txs: Transactions,
+    _uncle_list: Vec<Header>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Transactions {
+    tx_list: Vec<Transaction>,
+}
+
+impl Transactions {
+    pub fn root(&self) -> anyhow::Result<H256> {
+        let memdb = Arc::new(MemoryDB::new(true));
+        let mut trie = EthTrie::new(Arc::clone(&memdb));
+
+        // Insert txs into tx tree
+        for (index, tx) in self.tx_list.iter().enumerate() {
+            let path = rlp::encode(&index).freeze().to_vec();
+            let encoded_tx = tx.encode();
+            trie.insert(&path, &encoded_tx)
+                .map_err(|err| anyhow!("Error calculating transactions root: {err:?}"))?;
+        }
+
+        trie.root_hash()
+            .map_err(|err| anyhow!("Error calculating transactions root: {err:?}"))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Uncles {
+    uncle_list: Vec<Header>,
+}
+
+impl Uncles {
+    pub fn root(&self) -> anyhow::Result<H256> {
+        // generate rlp encoded list of uncles
+        let mut stream = RlpStream::new();
+        stream.append_list(&self.uncle_list);
+        let uncles_rlp = stream.out().freeze();
+
+        // hash rlp uncles
+        let mut keccak = Keccak256::new();
+        keccak.update(uncles_rlp);
+        let hash = keccak.finalize().to_vec();
```

Comment:
When there's just one chunk to hash, I think [`digest()`](https://docs.rs/sha3/0.9.1/sha3/trait.Digest.html#tymethod.digest) ought to work:
```suggestion
        let hash = Keccak256::digest(uncles_rlp);
```